### PR TITLE
Add simple in-memory Window function implementation

### PR DESCRIPTION
**Pull request by aditi-pandit**

### CODE REVIEW

**Comment by aditi-pandit** [comment]
Changes in progress : Basic window frame functionality.

File: velox/exec/Window.cpp [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+  }
+
+  numRowsApplied_ += numRows;
+  if (endRow == partitionStartRows_[currentPartitionIndex_ + 1]) {
+    currentPartitionIndex_++;
+  }
+}
+
+RowVectorPtr Window::getOutput() {
+  if (finished_ || !noMoreInput_ || returningRows_.size() == numRowsApplied_) {
```
Comment (mbasmanova) [inline velox/exec/Window.cpp:399]: Can we remove 'returningRows_.size() == numRowsApplied_' check by adding logic to set finished to true when we are done processing?

File: velox/exec/Window.cpp [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+  // ExtractColumn might be doing a copy. Try to avoid that.
+  for (int i = 0; i < inputColumnsSize_; ++i) {
+    data_->extractColumn(
+        returningRows_.data() + numRowsApplied_,
+        rowsForOutput,
+        i,
+        result->childAt(i));
+  }
+
+  std::vector<VectorPtr> windowFunctionOutputs;
```
Comment (mbasmanova) [inline velox/exec/Window.cpp:418]: No need for this vector. Use result instead.

File: velox/functions/prestosql/window/RowNumber.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    rowNumber_ = 1;
+  }
+
+  void apply(
+      const BufferPtr& peerGroupStarts,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& /*frameStarts*/,
+      const BufferPtr& /*frameEnds*/,
+      int32_t /*startPartitionRow*/,
+      int32_t resultIndex,
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/RowNumber.cpp:40]: perhaps, rename this to resultOffset or offset

File: velox/functions/prestosql/window/RowNumber.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  exec::registerWindowFunction(
+      name,
+      std::move(signatures),
+      [name](
+          const std::vector<exec::RowColumn>& argColumns,
+          const std::vector<TypePtr>& argTypes,
+          const TypePtr& /*resultType*/)
+          -> std::unique_ptr<exec::WindowFunction> {
+        VELOX_CHECK_EQ(argColumns.size(), 0, "{} takes no arguments", name);
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/RowNumber.cpp:67]: These checks are redundant as the framework already knows the signature and can do these checks in a generic manner. Let's remove them.

File: velox/functions/prestosql/window/RowNumber.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        VELOX_CHECK_EQ(argTypes.size(), 0, "{} takes no arguments", name);
+        return std::make_unique<RowNumberFunction>();
+      });
+  return true;
+}
+
+static bool FB_ANONYMOUS_VARIABLE(g_WindowFunction) =
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/RowNumber.cpp:74]: Aggregate functions are registered this way, but it is quite problematic. Let's follow the mechanism used for scalar functions which requires explicit register call.

File: velox/functions/prestosql/window/Rank.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      int32_t startPartitionRow,
+      int32_t resultIndex,
+      const VectorPtr& result) {
+    int numRows = peerGroupStarts->size();
+    auto* rawResultsVector =
+        result->asFlatVector<int64_t>()->mutableRawValues();
+    for (int i = 0; i < numRows; i++) {
+      if (peerGroupStarts->as<size_t>()[i] != currentPeerGroupStart_) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/Rank.cpp:48]: use temp variable for peerGroupStarts->as<size_t>()

File: velox/functions/prestosql/window/NthValue.cpp [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  explicit NthValueFunction(
+      const std::vector<exec::RowColumn>& argColumns,
+      const TypePtr& resultType)
+      : WindowFunction(resultType),
+        argColumns_(argColumns),
+        partitionRows_(nullptr) {}
+
+  void resetPartition(const std::vector<char*>& rows) {
+    partitionRows_ = reinterpret_cast<char* const*>(rows.data());
+    /*firstArgVector_->resize(rows.size());
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/NthValue.cpp:38]: remove commented out code

File: velox/functions/prestosql/window/NthValue.cpp [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+ public:
+  explicit NthValueFunction(
+      const std::vector<exec::RowColumn>& argColumns,
+      const TypePtr& resultType)
+      : WindowFunction(resultType),
+        argColumns_(argColumns),
+        partitionRows_(nullptr) {}
+
+  void resetPartition(const std::vector<char*>& rows) {
+    partitionRows_ = reinterpret_cast<char* const*>(rows.data());
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/NthValue.cpp:37]: Why do you need to reinterpret_cast?

File: velox/functions/prestosql/window/NthValue.cpp [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+  void apply(
+      const BufferPtr& peerGroupStarts,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& /*frameEnds*/,
+      int32_t startPartitionRow,
+      int32_t resultIndex,
+      const VectorPtr& result) {
+    FlatVector<T>* resultVector = result->as<FlatVector<T>>();
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/NthValue.cpp:55]: auto

File: velox/functions/prestosql/window/NthValue.cpp [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    auto firstArgVector =
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/NthValue.cpp:56]: re-use the vector across batches

File: velox/functions/prestosql/window/NthValue.cpp [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        BaseVector::create(resultVector->type(), 1, peerGroupStarts->pool());
+    auto offsetsVector =
+        BaseVector::create(BIGINT(), 1, peerGroupStarts->pool());
+
+    int numRows = peerGroupStarts->size();
+    for (int i = 0; i < numRows; i++) {
+      exec::RowContainer::extractColumn(
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/NthValue.cpp:63]: Why do we extract values one row at a time? Can we extract all rows at once before the loop?

File: velox/functions/prestosql/window/NthValue.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+          partitionRows_ + startPartitionRow + i,
+          1,
+          argColumns_[1],
+          offsetsVector);
+      const int64_t offset = offsetsVector->values()->as<int64_t>()[0];
+      VELOX_CHECK_GE(offset, 1);
+      const int64_t frameStart = frameStarts->as<size_t>()[i];
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/NthValue.cpp:70]: use temp variable for frameStarts->as<size_t>()

int64_t -> vector_size_t

File: velox/functions/prestosql/window/NthValue.cpp [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+      VELOX_CHECK_GE(frameStart, 0);
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/NthValue.cpp:71]: remove this check from the hot loop

File: velox/functions/prestosql/window/NthValue.cpp [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+      exec::RowContainer::extractColumn(
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/NthValue.cpp:72]: Let's modify RowContainer API to allow extracting the value directly onto resultVector.

File: velox/functions/prestosql/window/NthValue.cpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+          partitionRows_ + frameStart + offset - 1,
+          1,
+          argColumns_[0],
+          firstArgVector);
+      resultVector->mutableRawValues()[resultIndex + i] =
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/NthValue.cpp:77]: This won't work for strings as we need to copy not only StringView, but the underlying string as well.

File: velox/exec/WindowFunction.cpp [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+  if (it != functionsMap.end()) {
+    return &it->second;
+  }
+
+  return std::nullopt;
+}
+} // namespace
+
+bool registerWindowFunction(
+    const std::string& name,
```
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/exec/WindowFunction.cpp:41]: The name parameter is only used for lookup and not stored anywhere. So this needn't be done.

File: velox/exec/Window.cpp (hunk 1 above)
Comment (aditi-pandit) [reply velox/exec/Window.cpp:399]: That existed already. Removing this condition.

File: velox/exec/Window.cpp [hunk 17]
```diff
@@ -0,0 +1,465 @@
...
+  // We directly return above if there are no rows. So we can safely assume here
+  // that we have atleast 1 input row.
+  for (int64_t i = 1; i < returningRows_.size(); i++) {
+    if (partitionCompare(returningRows_[i - 1], returningRows_[i])) {
+      j++;
+      partitionStartRows_[j] = i;
+    }
+  }
+  currentPartitionIndex_ = 0;
+  numberOfPartitions_ = j + 1;
```
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/exec/Window.cpp:238]: I realized that this variable was unnecessary. Removed it.

File: velox/exec/WindowFunction.h [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+#include "velox/core/PlanNode.h"
+#include "velox/exec/RowContainer.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/BaseVector.h"
+
+namespace facebook::velox::exec {
+
+class WindowFunction {
+ public:
+  explicit WindowFunction(TypePtr resultType) : resultType_(resultType) {}
```
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/exec/WindowFunction.h:28]: So far I haven't encountered a function that needed input columns that weren't arguments. However, we can change this API if needed subsequently.

File: velox/exec/Window.cpp [hunk 19]
```diff
@@ -0,0 +1,465 @@
...
+    for (int w = 0; w < numFuncs; w++) {
+      // The peer and frame values set in the WindowFunction::apply buffers
+      // are the offsets within the current partition, so offset all row
+      // indexes correctly.
+      allFuncsRawPeerStartBuffer[w][j] = peerStartRow_ - firstPartitionRow;
+      allFuncsRawPeerEndBuffer[w][j] = peerEndRow_ - 1 - firstPartitionRow;
+
+      auto frameEndPoints = findFrameEndPoints(
+          w, partitionStartRows_[currentPartitionIndex_], endRow, i);
+      VELOX_CHECK_EQ(frameEndPoints.second, i);
```
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/exec/Window.cpp:373]: Removing this check. I'll also move all the checks in findFrameEndPoints to the constructor to fail early and reduce per-row logic.

File: velox/core/PlanNode.h [hunk 20]
```diff
@@ -57,6 +57,11 @@ class SortOrder {
   SortOrder(bool ascending, bool nullsFirst)
       : ascending_(ascending), nullsFirst_(nullsFirst) {}
 
+  SortOrder(const SortOrder& other) {
```
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/core/PlanNode.h:60]: This was needed for the initKeyInfo function (in Window.cpp) when adding the default sort order into the vectors.

File: velox/exec/Window.cpp (hunk 2 above)
Comment (aditi-pandit) [reply velox/exec/Window.cpp:418]: I have used this vector in intermediate computations.

File: velox/exec/WindowFunction.h [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+  // rows is a buffer of the all the rows in this partition
+  virtual void resetPartition(const std::vector<char*>& rows) = 0;
+
+  virtual void apply(
+      const BufferPtr& peerGroupStarts,
+      const BufferPtr& peerGroupEnds,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      int32_t startPartitionRow,
+      int32_t resultIndex,
```
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/exec/WindowFunction.h:45]: Yes, we can count the partition rows in the function implementation. I have removed that parameter.

File: velox/functions/prestosql/window/NthValue.cpp (hunk 14 above)
Comment (aditi-pandit) [reply velox/functions/prestosql/window/NthValue.cpp:72]: I am no longer using this API.

**Comment by aditi-pandit** [comment]
Squashed and cleaned the commits to https://github.com/facebookincubator/velox/pull/1816 for further reviews.

These review comments are also addressed in that PR. Lets continue further discussion there.

**Pull request closed without merging**